ding/model/template/ebm.py

```diff
@@ -0,0 +1,488 @@
+"""
+Vanilla DFO and EBM are adapted from https://github.com/kevinzakka/ibc.
+MCMC is adapted from https://github.com/google-research/ibc.
+"""
+from typing import Callable, Tuple
+
+import numpy as np
+import torch
+import torch.nn as nn
+import torch.nn.functional as F
+
+from abc import ABC, abstractmethod
+
+from ding.utils import MODEL_REGISTRY, STOCHASTIC_OPTIMIZER_REGISTRY
+from ding.torch_utils import unsqueeze_repeat
+from ding.model.wrapper import IModelWrapper
+from ..common import RegressionHead
+
+
+def create_stochastic_optimizer(device, stochastic_optimizer_config):
+    return STOCHASTIC_OPTIMIZER_REGISTRY.build(
+        stochastic_optimizer_config.pop("type"), device=device, **stochastic_optimizer_config
+    )
+
+
+def no_ebm_grad():
+    """Wrapper that disables energy based model gradients"""
+
+    def ebm_disable_grad_wrapper(func: Callable):
+
+        def wrapper(*args, **kwargs):
+            ebm = args[-1]
+            assert isinstance(ebm, (IModelWrapper, nn.Module)),\
+                   'Make sure ebm is the last positional arguments.'
+            ebm.requires_grad_(False)
+            result = func(*args, **kwargs)
+            ebm.requires_grad_(True)
+            return result
+
+        return wrapper
+
+    return ebm_disable_grad_wrapper
+
+
+class StochasticOptimizer(ABC):
+
+    def _sample(self, obs: torch.Tensor, num_samples: int) -> Tuple[torch.Tensor, torch.Tensor]:
+        """Helper method for drawing action samples from the uniform random distribution
+        and tiling observations to the same shape as action samples.
+        obs: (B, O)
+        return: (B, N, O), (B, N, A).
+        """
+        action_bounds = self.action_bounds.cpu().numpy()
+        size = (obs.shape[0], num_samples, action_bounds.shape[1])
+        action_samples = np.random.uniform(action_bounds[0, :], action_bounds[1, :], size=size)
+        action_samples = torch.as_tensor(action_samples, dtype=torch.float32).to(self.device)
+        tiled_obs = unsqueeze_repeat(obs, num_samples, 1)
+        return tiled_obs, action_samples
+
+    @staticmethod
+    @torch.no_grad()
+    def _get_best_action_sample(obs: torch.Tensor, action_samples: torch.Tensor, ebm: nn.Module):
+        """Return target with highest probability (lowest energy).
+        obs: (B, N, O), action_samples: (B, N, A)
+        return: (B, A).
+        """
+        # (B, N)
+        energies = ebm.forward(obs, action_samples)
+        probs = F.softmax(-1.0 * energies, dim=-1)
+        # (B, )
+        best_idxs = probs.argmax(dim=-1)
+        return action_samples[torch.arange(action_samples.size(0)), best_idxs]
+
+    def set_action_bounds(self, action_bounds: np.ndarray):
```

Comment:
maybe implement it as property and its setter

Reply:
changed